[feature] ArrowKeys to navigate between image results

BernieHuang2008 opened the pull request:
## What does this PR do?

<!-- MANDATORY -->

<!-- explain the changes in your PR, algorithms, design, architecture -->
Improved PR #2721.
## Why is this change important?

<!-- MANDATORY -->

<!-- explain the motivation behind your PR -->

## How to test this PR locally?

<!-- commands to run the tests or instructions to test the changes-->

## Author's checklist
I use Switch-case to make the code more readable and flexible, it's completely same as the example @return42 gived in #2721.

<!-- additional notes for reviewiers -->

## Related issues
Closes #2720
Closes #2736
<!--
Closes #234
-->

BernieHuang2008 committed:
[feature] key bindings: left & right arrow are always on

bonswouar commented on searx/static/themes/simple/src/js/main/keyboard.js:
```diff
@@ -43,6 +43,29 @@ searxng.ready(function () {
+        searxng.selectPrevious();
+        break;
+      case 'ArrowRight':
+        searxng.selectPrevious();
```
Typo, should be `searxng.selectNext();` :)

BernieHuang2008 replied:
thanks, solved.

bonswouar commented on searx/static/themes/simple/src/js/main/keyboard.js:
```diff
@@ -43,6 +43,29 @@ searxng.ready(function () {
     }
   });
 
+  searxng.on('.result', 'keydown', function (e) {
```
Why not use `.image-detail-open` class directly then?
Plus with that the next 2 lines can be removed (no check for image result needed)

BernieHuang2008 replied:
that all works same, because `.image-detail-open` contains `.result`.

BernieHuang2008 replied:
However, it seems has a new problem:

It's about the focus point: 
If user focus on the image (e.g., click the image), we won't catch the event when binding on `'.result'`
Otherwise, if we bind it on `.detail`(which is the pannel that display the image) but the user didn't do anything, then the focus is still on `.result`, we can't catch it on `.detail`.

BernieHuang2008 replied:
maybe we can bind it on `body` and use your `hasImageInDetail` to check whether a image result is opened.

```js
  function hasImageInDetail () {
    return document.querySelector('#results.image-detail-open') !== null;
  }
```
what do you think?

bonswouar replied:
> that all works same, because .image-detail-open contains .result.

Yes but that should be cleaner, as the binding is done only when required (when an image is selected), not when any result is shown.

> It's about the focus point: If user focus on the image (e.g., click the image), we won't catch the event when binding on `'.result'` Otherwise, if we bind it on `.detail`(which is the pannel that display the image) but the user didn't do anything, then the focus is still on `.result`, we can't catch it on `.detail`.

Not sure what you mean exactly, but in my tests locally just binding on `.image-detail-open` seems to work properly.
But yes the user needs to click on one image first (resulting in showing `.detail` and adding `.image-detail-open`), that's the expected behaviour, isn't it?

BernieHuang2008 replied:
for the first comment, i agree with you. (committed)
for the second, i think i should make a video to explain this:

https://github.com/searxng/searxng/assets/88757735/efebe378-fde2-4547-8d43-e3fbc48a05ec

You see, that's the problem and we should solve it.

bonswouar replied:
What are the steps? On my end that seems to work as intended, I click on an image and then use the arrow keys. Maybe I missed something?
I don't have any condition on `e.target` locally though, as said in the reviews

BernieHuang2008 replied:
well, it's really hard to explain, so i think we should put it away right now. 
We can solve it later ?

BernieHuang2008 replied:
> What are the steps? On my end that seems to work as intended, I click on an image and then use the arrow keys. Maybe I missed something? I don't have any condition on `e.target` locally though, as said in the reviews

maybe you missed the step of clicking the blank when the image preview is opened.

BernieHuang2008 replied:
i think we can start a new issue or discussion to talk about this. how about #2736 ?

bonswouar commented on searx/static/themes/simple/src/js/main/keyboard.js:
```diff
@@ -43,7 +43,7 @@ searxng.ready(function () {
-  searxng.on('.result', 'keydown', function (e) {
+  searxng.on('.image-detail-open', 'keydown', function (e) {
     var resultElement = getResultElement(e.target);
     if (isImageResult(resultElement)) {
```
If you go with the binding on `.image-detail-open` then this condition `if (isImageResult(resultElement))` isn't required anymore (`.image-detail-open` only exists when an image is opened) as said previously

dalf commented on searx/static/themes/simple/src/js/main/keyboard.js:
```diff
@@ -43,6 +43,26 @@ searxng.ready(function () {
     }
   });
 
+  searxng.on('.image-detail-open', 'keydown', function (e) {
```
Currently `Ctrl-Left`, `Alt-Left` etc... works too.

Can you add something like this:

```js
if (e.ctrlKey || e.altKey || e.shiftKey || e.metaKey) {
  return
}
```

to ignore the event when Ctrl/Alt/Shift/Meta keys are pressed ?

BernieHuang2008 replied:
Committed, are there anything else I can do?

return42 committed:
[mod] improve handling of the hotkeys

- KeyboardEvent: keyCode property is depricated, replaced by key property [2]

- the check for ifDetailOpened is not necessary, because the hotkeys are not
  only applicable to image-results, by example:

   `!goi !go !scc hello`

- Key bindings like h for help are to be used in general (not only in vim-mode)

[1] https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/keyCode
[2] https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/key

Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>

return42 committed:
[build] /static
